pennylane/operation.py

```diff
@@ -2978,6 +3002,19 @@ def disable_new_opmath():
     global __use_new_opmath
     __use_new_opmath = False
 
+    if _mock_ham_stack:
+        _mock_ham_stack.pop().close()
+    if _mock_lc_stack:
+        return
+
+    with contextlib.ExitStack() as stack:
+        try:
+            _mock = mock.patch("__main__.Hamiltonian", qml.Hamiltonian)
+            stack.enter_context(_mock)
+            _mock_lc_stack.append(stack.pop_all())
```

Comment:
```suggestion
            from __main__ import Hamiltonian as MainHamiltonian

            if MainHamiltonian is qml.ops.LinearCombination:
                _mock = mock.patch("__main__.Hamiltonian", qml.Hamiltonian)
                stack.enter_context(_mock)
                _mock_lc_stack.append(stack.pop_all())
```
I also realized that we should only mock the user's `Hamiltonian` if it actually is a `LinearCombination`, otherwise we risk overwriting an unrelated user variable.

I appreciate that this is getting increasingly hacky. Would be good to bring in @mlxd here for thoughts, but I'm open to considering the non-patching of `from pennylane import Hamiltonian` as an edge case that doesn't need to be resolved in this PR.

Reply:
> non-patching of from pennylane import Hamiltonian as an edge case that doesn't need to be resolved in this PR.

I'd be okay with that 👍

Reply:
> > non-patching of from pennylane import Hamiltonian as an edge case that doesn't need to be resolved in this PR.
>
> I'd be okay with that 👍

Ok, let's not worry about that right now.

Reply:
Okay so lets revert the change?

Reply:
Yes please!

Reply:
Done ✅ I reset to the commit before Mudits update